bpf/include/bpf/builtins.h

```diff
@@ -6,17 +6,240 @@
+	case 48: jmp_48: __it_mov(d, s, 64);
+	case 40: jmp_40: __it_mov(d, s, 64);
+	case 32: jmp_32: __it_mov(d, s, 64);
+	case 24: jmp_24: __it_mov(d, s, 64);
+	case 16: jmp_16: __it_mov(d, s, 64);
+	case  8: jmp_8:  __it_mov(d, s, 64);
+		break;
+
+	case 94: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_88;
+	case 86: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_80;
+	case 78: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_72;
+	case 70: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_64;
+	case 62: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_56;
+	case 54: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_48;
+	case 46: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_40;
+	case 38: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_32;
+	case 30: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_24;
+	case 22: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_16;
+	case 14: __it_mov(d, s, 16); __it_mov(d, s, 32); goto jmp_8;
+	case  6: __it_mov(d, s, 16); __it_mov(d, s, 32);
+		 break;
+
+	case 92: __it_mov(d, s, 32); goto jmp_88;
+	case 84: __it_mov(d, s, 32); goto jmp_80;
+	case 76: __it_mov(d, s, 32); goto jmp_72;
+	case 68: __it_mov(d, s, 32); goto jmp_64;
+	case 60: __it_mov(d, s, 32); goto jmp_56;
+	case 52: __it_mov(d, s, 32); goto jmp_48;
+	case 44: __it_mov(d, s, 32); goto jmp_40;
+	case 36: __it_mov(d, s, 32); goto jmp_32;
+	case 28: __it_mov(d, s, 32); goto jmp_24;
+	case 20: __it_mov(d, s, 32); goto jmp_16;
+	case 12: __it_mov(d, s, 32); goto jmp_8;
+	case  4: __it_mov(d, s, 32);
+		 break;
+
+	case 90: __it_mov(d, s, 16); goto jmp_88;
+	case 82: __it_mov(d, s, 16); goto jmp_80;
+	case 74: __it_mov(d, s, 16); goto jmp_72;
+	case 66: __it_mov(d, s, 16); goto jmp_64;
+	case 58: __it_mov(d, s, 16); goto jmp_56;
+	case 50: __it_mov(d, s, 16); goto jmp_48;
+	case 42: __it_mov(d, s, 16); goto jmp_40;
+	case 34: __it_mov(d, s, 16); goto jmp_32;
+	case 26: __it_mov(d, s, 16); goto jmp_24;
+	case 18: __it_mov(d, s, 16); goto jmp_16;
+	case 10: __it_mov(d, s, 16); goto jmp_8;
+	case  2: __it_mov(d, s, 16);
+		 break;
+
+	case  1: __it_mov(d, s, 8);
+		break;
+
+	default:
+		/* __builtin_memcpy() is crappy slow since it cannot
+		 * make any assumptions about alignment & underlying
+		 * efficient unaligned access on the target we're
+		 * running.
+		 */
+		__throw_build_bug();
+	}
+#else
+	__bpf_memcpy_builtin(d, s, len);
 #endif
+}
+
+static __always_inline __maybe_unused void
+__bpf_no_builtin_memcpy(void *d __maybe_unused, const void *s __maybe_unused,
+			__u64 len __maybe_unused)
+{
+	__throw_build_bug();
+}
+
+/* Unfortunately verifier forces aligned stack access while other
+ * pointers do not have to be aligned (map, pkt, etc). Mark those
+ * on the stack for objects > 8 bytes in order to force-align such
+ * memcpy candidates when we really need them. Usually not needed
+ * for <= 8 bytes and in case of > 8 bytes only when 8 byte is not
+ * the natural alignment.
+ */
+#define __align_stack_8		__aligned(8)
```

Comment:
Would be nice to have some static test that we are using this macro everywhere we should., but I'm not sure I understand when we should use it. For all on-stack objects of size >8B? Only for those that get used in `memcpy`/`memset`? This commit seems to do neither, so I'm probably missing something.

Reply:
On stack objects >8b that are passed in either memcpy/memset and have a natural alignment of < 8b, e.g. `__u8 foo[12]`. Turns out we don't have that many objects with such property, so it's not needed often.

Reply:
Would be nice to have a static check for that. Not sure Coccinelle is the best fit to do that (I don't think it can match on objects' sizes and alignments), but I can have a look. Worth-case scenario, we hardcode the known-problematic structures in the semantic patch.

Reply:
My smpl skills are only very rusty, I tried coming up with something yesterday night, but got stuck here basically. I'm probably looking at it from the wrong angle in that the below might not be possible this way as in finding the declaration of the expression `T *x` we pass in below and adding the attribute to it. I got stuck at the mix of expression from the match and getting the identifier for m.x:
```
@m@
expression e1, e2, e3, e4;
type T;
T *x;
@@
(
        ctx_load_bytes(e1, e2, x, e3);
|
        ctx_store_bytes(e1, e2, x, e3, e4);
)

@@
attribute name __align_stack;
//expression m.x
identifier x;
type T;
@@
-       T x;
+       T x __align_stack;
```
If you find something I would be very curious! :-)

Reply:
Agree that it might be easier finding the bad cases explicitly by name though. Also, for most cases in our code base things will just work due to natural object alignment, so it's really just a few cases luckily.